### PR TITLE
io_uring: document public interface, limit visibility and rework unsafe usage

### PR DESCRIPTION
**Pull request by alindima**
# Reason for This PR

Fixes https://github.com/firecracker-microvm/firecracker/issues/2759

Also adds some other minor refactoring: see the commits

## Description of Changes

See the commits

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: src/devices/src/virtio/block/io/async_io.rs [hunk 1]
```diff
@@ -118,18 +118,21 @@ impl<T> AsyncFileEngine<T> {
 
         let wrapped_user_data = WrappedUserData::new_with_dirty_tracking(addr, user_data);
 
-        self.ring
-            .push(Operation::read(
+        // Safe because we trust that the host kernel will us pass back a completed entry with this
```
Comment (serban300) [inline src/devices/src/virtio/block/io/async_io.rs:121]: Seems like there's a topic issue: `will us pass back`. It's present in other comments as well.

File: src/io_uring/src/lib.rs [hunk 2]
```diff
@@ -63,6 +76,7 @@ pub enum Error {
 }
 
 impl Error {
+    /// Return true if this is error is caused by a full submission queue.
```
Comment (serban300) [inline src/io_uring/src/lib.rs:79]: Did you mean `if this error is caused ...` ?

File: src/io_uring/src/operation/sqe.rs [hunk 3]
```diff
@@ -7,17 +7,21 @@ use crate::bindings::io_uring_sqe;
 
 unsafe impl ByteValued for io_uring_sqe {}
 
-pub struct Sqe(pub(crate) io_uring_sqe);
+/// Newtype wrapper over a raw sqe.
```
Comment (serban300) [inline src/io_uring/src/operation/sqe.rs:10]: Did you mean `New type` ?

File: src/devices/src/virtio/block/io/async_io.rs (hunk 1 above)
Comment (alindima) [reply src/devices/src/virtio/block/io/async_io.rs:121]: done, copy paste error 😄

File: src/io_uring/src/operation/sqe.rs (hunk 3 above)
Comment (alindima) [reply src/io_uring/src/operation/sqe.rs:10]: Newtype is actually a small Rust design pattern: https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html

File: src/io_uring/src/lib.rs (hunk 2 above)
Comment (alindima) [reply src/io_uring/src/lib.rs]: yep, done

**Review by serban300 — DISMISSED** [review]

**Review by serban300 — APPROVED** [review]

**Review by alsrdn — APPROVED** [review]

**Review by serban300 — APPROVED** [review]

**Pull request merged**